cirq/contrib/two_qubit_gates/gate_compilation.py

```diff
@@ -0,0 +1,328 @@
+            within some Pauli error.
+        base_gate: The base 2 qubit gate used in the gate product.
+        max_dist: The largest allowed Pauli error between a generated 2Q
+           unitary and a KAK vector mesh point that it is tabulated to.
+        KAK_mesh: Sequence of KAK vectors filling the Weyl chamber whose
+            nearest neighbor distance is about 2*max_error.
+        *local_unitary_pairs: 2-Tuples of single qubit unitary tensors, each
+            of shape (N,2,2).
+
+    Returns:
+        kept_kaks: Generated KAK vectors that are uniquely close to at least one
+            mesh point.
+        kept_cycles: The corresponding single qubit unitaries required to obtain
+            the desired KAK vectors.
+    """
+    shapes = {pair[0].shape for pair in local_unitary_pairs}
+    shapes.update({pair[0].shape for pair in local_unitary_pairs})
+    assert len(shapes) == 1
+    assert len(shapes.pop()) == 3
+
+    # Generate products
+    local_cycles = np.array(
+        [vector_kron(*pairs) for pairs in local_unitary_pairs])
+
+    prods = np.einsum('ab,...bc,cd', base_gate, local_cycles[0], base_gate)
+    for local_cycle in local_cycles[1:]:
+        np.einsum('ab,...bc,...cd', base_gate, local_cycle, prods, out=prods)
+
+    kak_vectors = kak_vector(prods, check_preconditions=False)
+
+    kept_kaks = []
+    kept_cycles = []
+
+    for ind, vec in enumerate(kak_vectors):
+        # dists = KAK_vector_infidelity(vec, KAK_mesh)
+        # The L2 distance is an upper bound to the locally invariant distance,
+        # but it's much faster to compute.
+        dists = np.sqrt(np.sum((KAK_mesh - vec)**2, axis=-1))
+        close = (dists < max_dist).nonzero()[0]
+        assert close.shape[0] in (0, 1), f'shape: {close.shape}'
+        cycles_for_gate = tuple(
+            (k_0[ind], k_1[ind]) for k_0, k_1 in local_unitary_pairs)
+
+        kept = False
+        for mesh_ind in close:
+            if mesh_ind not in tabulation:
+                tabulation[mesh_ind] = cycles_for_gate
+                kept = True
+
+        if kept:
+            kept_kaks.append(vec)
+            kept_cycles.append(cycles_for_gate)
+
+    return kept_kaks, kept_cycles
+
+
+def gate_product_tabulation(base_gate: np.ndarray,
+                            max_infidelity: float,
+                            verbose: bool = False,
+                            include_warnings: bool = True) -> GateTabulation:
+    r"""Generate a GateTabulation for a base two qubit unitary.
+
+    Args:
+        base_gate: The base gate of the tabulation.
+        max_infidelity: Sets the desired density of tabulated product unitaries.
+            The typical nearest neighbor Euclidean spacing (of the KAK vectors)
+            will be on the order of \sqrt(max_infidelity). Thus the number of
+            tabulated points will scale as max_infidelity^{-3/2}.
+        verbose: Whether to print gate tabulation statistics.
```

Comment:
Remove the verbose parameter. Replace it with methods on `GateTabulation` that provide access to the information that would have been printed.

Reply:
Done.